### ISSUE TITLE
didInsert called with null node.el when mount() is followed by immediate redraw()

### ISSUE THREAD

**Issue by talmobi** [leeoniya/domvm#74]
Hi,

I'm seeing didInsert being called twice - once with the node and once without. So you can effectively ignore the faulty one (it IS faulty and a bug, right?) by simply doing "if (!node) return"

I have an Dashboard view with Chart views. The chart views have didInsert hooks that are called twice.

Something like this:
```js
var insert_counter = 1 // didInsert counter - is called 6 times
var nodes_counter = 1 // didInsert with valid node's - called 3 times

function Chart (vm, opts) {
  function didInsert (node) {
    console.log("counter: " + insert_counter++)
    if (!node) return
    console.log("valid nodes counter: " + nodes_counter++)
  }
  return ['div.chart',
    ['canvas', {_hooks: {didInsert: didInsert}}]
  ]
}

function Dashboard (vm, store) {
  return function () {
    ['div.dashboard',
      ['div', [Chart, opts]],
      ['div', [Chart, opts]],
      ['div', [Chart, opts]]
    ]
  }
}
```

**Comment by talmobi** [comment]
Ok ok, I think I know what's going on - not sure if this is a bug or no though? Better way?

So I have the domvm router in use - and on every path it has the "onenter" callback attached and using "vm.redraw()".

So on the initial page load, the redraw is being called twice. If I remove it, the initial page load works as expected (no double didInsert with empty nodes). But then the routing of course won't work.

So I'm not sure how I can detect if the page is on initial render mode and ignore the vm.redraw?

This is what the router looks like:

```
  router.config({
    useHist: false,
    root: '/',
    init: function () {
      vm.mount(document.body)
    }
  })

  return {
    home: {
      path: '/',
      onenter: function () {
        vm.redraw()
        console.log('onenter')
      },
      onexit: function () {}
    },
    about: {
      path: '/about',
      onenter: function () {
        vm.redraw()
        console.log('onenter')
      },
      onexit: function () {}
    },
    dashboard: {
      path: '/dashboard',
      onenter: function () {
        vm.redraw()
        console.log('onenter')
      },
      onexit: function () {}
    }
  }

```

**Comment by talmobi** [comment]
Hi again, ok so at the end of my router I have (like the examples):

```
var deps = {}
var router = domvm.route(Router, deps)
router.refresh()
```

So what's happening is that router.refresh() and the routers vm.redraw() are both called. The second call is called with nodes being null.

It seems they are kind of mutual exclusive since you can call "vm.redraw()" multiple times with no effect, as well as router.refresh() multiple times with no effect. But using them both will nullify the other (hooks being called with null nodes).

So the fix for me was to basically comment out router.refresh() and all works as expected:

```
var deps = {}
var router = domvm.route(Router, deps)
//router.refresh()
```

**Comment by leeoniya** [comment]
@talmobi 

A full example that triggers the behavior would be more helpful. For instance in `[Chart, opts]` is `opts` the same object for each chart or opts1/opts2/opts3? Can you make a minimal but working demo of what you're seeing? Also make sure you're testing with the `1.x-dev` branch, it had a routing issue (https://github.com/leeoniya/domvm/issues/73) pushed yesterday but i don't think it's related. The following works as expected for me:

```js
var insert_counter = 0;
var nodes_counter = 0;

function Chart (vm, opts) {
	function didInsert (node) {
		console.log("counter: " + ++insert_counter)
		if (node)
			console.log("valid nodes counter: " + ++nodes_counter)
	}
	return function() {
		return ['.chart',
			['canvas', {_hooks: {didInsert: didInsert}}]
		];
	}
}

function Dashboard (vm, store) {
	var opts1 = {};
	var opts2 = {};
	var opts3 = {};

	return function () {
		return ['.dashboard',
			['div', [Chart, opts1]],
			['div', [Chart, opts2]],
			['div', [Chart, opts3]],
		];
	}
}

function Router(router, deps) {
	var vm = deps.vm;

	router.config({
		useHist: false,
		root: '/',
		init: function () {
			vm.mount(document.body)
		}
	})

	return {
		home: {
			path: '/',
			onenter: function () {
				vm.redraw()
				console.log('onenter')
			}
		}
	}
}

var deps = {
	vm: domvm.view(Dashboard, {})
};

var router = domvm.route(Router, deps);

router.refresh();
```

**Comment by talmobi** [comment]
Hi,

Yeah i was going to make a sample but sth else came up at work. I'll make one tonight and commed on this thhread.

**Comment by talmobi** [comment]
Hello, I made a working sample of the issue at:

https://github.com/talmobi/domvm_issue74/blob/master/index.html

you have to clone it and open index.html (and open console). I added two TODO tags (on line 30 and 147) where doing either of what they say fixes the issue.

Basically the issue seems to be that didInsert is being called twice, once with a node.el and once without. Uncommenting the router.refresh seems to fix the issue as well as checking for node.el === null inside the didInsert callback.

You with better insight of what's going on and what's causing the issue - perhaps I'm using the library wrong?

I'm not familiar enough with the lib at the moment to know what and how to fix, or even if it's a legitimate issue or no, so therefore can't issue a pull request.

**Comment by talmobi** [comment]
For example if you leave router.refresh() in and resize the window, the window resize events will fire (that were added inside the didInsert callback (or more accurately the initRealtimeChart callback that is being called by didInsert). The resize event will fire 8 times, 4 times with success, and 4 times failure because canvas (node.el) is null.

If you comment out router.refresh the resize event fires 4 times only all with success.

So seems an additional 4 nodes with null node.el are added somewhere (not visible in the DOM though - probably since node.el is null).

If you comment out the routers.onenter function that calls vm.redraw and leave in router.refresh it also works perfectly without issues (routing won't work of course) - but seems like when you use them both in the initial render it causes the issue (if you go to another route and go back to the dashboard route it will work fine (probably since router.refresh only comes into play in the very first render).

So a simple fix might be to simply not call router.refresh?

**Comment by leeoniya** [comment]
thanks for the demo. i had to cut out a lot of code that was unrelated to the issue. i'll post code examples tomorrow and explain. it's not a bug though and also not related to routing but the order in which mount() and redraw() get called.

**Comment by leeoniya** [comment]
So there are a couple things going on that cause this in the example. The fact that `didInsert` is invoked twice is a misunderstanding with how domvm treats models (explained below).

1. Model recreation on every redraw (the charts' `opts` objects are the models in this case). domvm defaults to assuming that models passed into views are persistent across redraws; if they aren't, the views will get destroyed and recreated as the models objects are regenerated. There are 3 ways you can solve this:

a) create an opt array outside of render() so the opt objects are the same ones used on each redraw.
```js
function Dashboard (vm) {
	var opts = [
		{ interval: 2500,  line_width: 3 },
		{ interval: 5000,  line_width: 8 },
		{ interval: 60000, line_width: 2 },
		{ interval: 20000, line_width: 1 },
	];

	return function () {
		return ['#dash',
			[Chart, opts[0]],
			[Chart, opts[1]],
			[Chart, opts[2]],
			[Chart, opts[3]],
		]
	}
}
```
b) provide persistent keys for the models, if you're creating them ad-hoc/in place within render():

```js
function Dashboard (vm) {
	return function () {
		return ['#dash',
			[Chart, { interval: 2500,  line_width: 3 }, 0],
			[Chart, { interval: 5000,  line_width: 8 }, 1],
			[Chart, { interval: 60000, line_width: 2 }, 2],
			[Chart, { interval: 20000, line_width: 1 }, 3],
		]
	}
}
```
c) similar to b), provide `false` for keys. This opts the view out of model persistance assumptions, essentially turning them into slots with swappable models (there's a bit more nuance here if you're actually swapping out different model objects, but ask if you're interested).

```js
function Dashboard (vm) {
	return function () {
		return ['#dash',
			[Chart, { interval: 2500,  line_width: 3 }, false],
			[Chart, { interval: 5000,  line_width: 8 }, false],
			[Chart, { interval: 60000, line_width: 2 }, false],
			[Chart, { interval: 20000, line_width: 1 }, false],
		]
	}
}
```

This takes care of the `didInsert` hooks being invoked twice. Once this is corrected through, the `didInsert` hooks get called when `node.el = null`. This looks like a bug which I suspect is caused by `redraw()` coalescing and immediately invoking `vm.redraw` after `vm.mount`. You can work around this by putting `router.refresh()` (but technically `vm.redraw`) within a `setTimeout/0`.

```js
function App (vm) {
	return function () {
		return ['#app', [Dashboard]];
	}
}

function Dashboard (vm) {
	return function () {
		return ['#dash',
			[Chart, { interval: 2500,  line_width: 3 }, 0],
			[Chart, { interval: 5000,  line_width: 8 }, 1],
			[Chart, { interval: 60000, line_width: 2 }, 2],
			[Chart, { interval: 20000, line_width: 1 }, 3],
		]
	}
}

function Chart (vm) {
	function didInsert (node) {
		console.log("node.el:", node.el);
	}

	return function () {
		return ['canvas.chart', {_hooks: { didInsert: didInsert }}]
	}
}

var vm = domvm.view(App).mount(document.body)

//setTimeout(function() {
	vm.redraw();
//}, 0)
```

I'll update this once I dig a bit more into it, however since it looks like you're not injecting your router into your view, you may consider a bit of simplification to your structure that avoids this issue. The main thing that's responsible for correct behavior here is that the initial subView is not identical to the subview set by `router.refresh()` and subsequent redraw() invocation. If you declare the initial subview as `var subView = [Dashboard];`, the issue comes back.

```js
var subView = null;
var appView = domvm.view(App).mount(document.body);
var router = domvm.route(Router);

router.refresh();

function Router(router) {
	return {
		dashboard: {
			path: '/',
			onenter: function () {
				subView = [Dashboard];
				appView.redraw();
			}
		}
	}
}

function App(vm) {
	return function () {
		return ['#app',
			subView || "Loading..."
		]
	}
}

function Dashboard (vm) {
	return function () {
		return ['#dashboard',
			[Chart, { interval: 2500,  line_width: 3 }, 0],
			[Chart, { interval: 5000,  line_width: 8 }, 1],
			[Chart, { interval: 60000, line_width: 2 }, 2],
			[Chart, { interval: 20000, line_width: 1 }, 3],
		]
	}
}

function Chart (vm) {
	function didInsert(node) {
		console.log("node.el:", node.el);
	}

	return function () {
		return ['canvas.chart', {_hooks: { didInsert: didInsert }}]
	}
}
```

**Comment by talmobi** [comment]
Thank you very much for your in-depth explanation. I see you've changed the title to better name the bug that may occur, so I'll leave the issue open for now - feel free to close it at any time you deem appropriate.

Will probably use domvm in a project of mine once I get the chance to dump legacy atrocities. Very nice work, thanks again!

**Comment by leeoniya** [comment]
@talmobi 

another workaround you can consider if you're interested in the root node of the vm is using `vm.hook`, which seems to be immune to this issue.

```js
function Chart(vm) {
	vm.hook({
		didMount: function(vm) {
			console.log("mounted node.el:", vm.node.el);
		}
	})

	return function () {
		return ['canvas.chart']
	}
}
```

**Comment by leeoniya** [comment]
here's a minimal example that triggers the issue:

```js
function View() {
	function didInsert(node) {
		console.log("node.el:", node.el);
	}

	return function () {
		return ['div', {_hooks: { didInsert: didInsert }}]
	}
}

var vm = domvm.view(View).mount(document.body);

vm.redraw();
```

**Comment by leeoniya** [comment]
So turns out this is a bit of a thorny issue. Here are some details.

`vm.redraw` is debounced/coalesced so if you call it 20 times within the same `requestAnimationFrame`, it will only fire the last invocation on the tail end of the frame. In addition, `did*` hook firing is delayed by 1-2 frames to ensure that the created DOM elements are "live" and are useful to you.

What's happening here is that the queue is filled with the callbacks and *current* nodes/args via the implicit `redraw()` as part of `mount()`, but the `did*` hooks don't fire until the next frame. If you call another `redraw()` before the queue is drained, those nodes in the queue become stale because they are regenerated on each redraw and the `el` transferred/recycled to the new vnodes. So when the hooks *do* fire, you're getting stale nodes that should be garbage collected.

What I'm gonna do is have domvm queue any `redraw()` calls if the `didHooks` queue is non-empty to prevent node replacement/recycling prior to hook invocation. Since the `didHooks` queue (and now `pendRedraws` queue) is global and cannot really be contained to a specific view, the solution is not a general one and may cause odd interactions in cases of significant complexity or asynchronicity WRT implicit sub-sub-subview redraw containing deep `did*` hooks cases. But it should cover multiple common cases like explicit view redraw like the one here.

**Comment by leeoniya** [comment]
@talmobi can this be closed?

**Comment by lawrence-dol** [comment]
> those nodes in the queue become stale because they are regenerated on each redraw

This is a general problem for all vDOM systems where redraws can cause wholesale replacement of nodes with identical or near-identical counterparts. The extent of the impact depends a lot on whether the UI component event queue is the same as the JavaScript event queue. If the UI event handling simply drops events on a secondary event queue there's an inescapable race-condition between UI and event handling. Due to CSS styling and cross-system difficulties it's likely that all browsers use a drawing canvas and handle their own events synthetically from events generated on the canvas; this is what I had to do with our applet and the reason for creating a full GUI toolkit on top of AWT's Canvas component.

There's an inherent race-condition between events and `redraws`. If several events happen in quick succession and are queued, and the first event does an inline redraw, it can and will invalidate the references for the remaining already enqueued events.

However, if a single queue is used and redraws are always deferred to the event queue, then that race-condition goes away since the redraws for all events in the queue are themselves coalesced and a single redraw event is queued up behind any already pending events, leaving their nodes intact until after they've been processed. And since O/S events remain pending on the GUI event queue (for the `Canvas` en toto), there's never a race between input components and their events.

But there is a lot on implementation dependent detail in what I just said; it depends on how composite/synthetic events are generated by the system. For example, take KeyDown, KeyTyped. Since the KeyTyped events depend on the event handling of KeyDown (e.g. they are only generated if the KeyDown preventDefault is not used), they must be queued discreetly apart from KeyDown. Therefore if KeyDown does an inline recreation of the input element, then KeyTyped events may be lost if the input component is replaced after they are generated but before they are processed.

With all that said, I've been deferring redraws to the event queue in both Mithril and DOMVM and have never had problems with input events being lost, though I did with inline draws with Mithril early on.

The crux of the problem is replacing input nodes when they've only been updated (e.g. as a result of key input changing the value and then replacing the component because it's value is changed); recycling becomes vital for such DOM nodes. DOMVM needs to ensure that instead of logically doing the equivalent of `new InputField()` it instead does `input.setValue(...)`.

**Comment by lawrence-dol** [comment]
PS: The use of RAF may mask but not eliminate the problem due to differences in granularity of handling several queues within JavaScript. The RAF queue is intended to be finer grained (more frequent) than the UI event queue.

**Comment by leeoniya** [comment]
> DOMVM needs to ensure that instead of logically doing the equivalent of new InputField() it instead does input.setValue(...)

The problem is that there is a lot more overhead in determining if two versions of things are identical vs at all different. Immutability is one solution but without language-level support, it just shifts this diff cost from the vdom lib to the immut lib. Performance would suffer if the default was isSame() rather than isDiff(). It could be something to explore but when you need such granular control the benefits of declarative start to diminish.

I'm not convinced there's a fundamental need for re-work here, but if anyone wants to explore replacing the core diff/transplanation logic of the lib while passing all tests and maintaining perf, I would be happy to look at that pull request.

**Comment by lawrence-dol** [comment]
> I'm not convinced there's a fundamental need for re-work here

Nor am I. I was just hoping to shed more light on the general problem in case it was helpful. It does not affect me, since I am already deferring and coalescing redraws, and any user of DOMVM can do the same in a few lines of code. But, just in case more things like this come out of the woodwork, I thought this might be useful information.

**Comment by leeoniya** [comment]
@talmobi 

https://github.com/leeoniya/domvm/commit/3f9d1d5313f8e9d62b3431ff8925ea55161267e3 significantly changed the internals of how `did*` fire, and removed the pending redraw queue. I tested to make sure that https://github.com/leeoniya/domvm/issues/74#issuecomment-219238966 still functioned properly and it seems to be okay.

could you verify that the current 1.x-dev version does not break anything for you?

thanks!

**Comment by talmobi** [comment]
@leeoniya 

Hi! The didInsert hooks are no longer been called with null nodes! So that's good.

Has router.refresh() been changed?

Calling

```
var router = domvm.route(Router, {});
router.refresh()
```

will call didInsert twice and subsequent calls in a setTimeout, e.g.:

```
    setTimeout(function () {
      console.log('calling router refresh')
      router.refresh()
    }, 5000)

```

will make the didInsert calls trigger again -- this is intentional?

**Comment by talmobi** [comment]
Basically this is no longer an issue:

https://github.com/talmobi/domvm_issue74/blob/master/index.html#L31-L32

But that also means that this fix is now mandatory:

https://github.com/talmobi/domvm_issue74/blob/master/index.html#L147-L148

But perhaps the latter never was an issue and I was just doing it wrong?

This code will create multiple didInsert calls, but that's perhaps the intended functionality?
https://github.com/talmobi/domvm_issue74/blob/master/14_6_2016.index.html#L144-L160

**Comment by leeoniya** [comment]
What specifically do what to accomplish by calling `router.refresh()`? There is potentially a lot of simplification that can be made here in general. I do think you're doing a couple things wrong but without knowing exactly what you want, I can't tell you how to get to where you're going.

You should not need any hooks at all (neither didInsert, nor didUnmount) in this example at all since you can directly insert raw DOM nodes as children into domvm templates. The reason you have multiple didInsert calls is not related to the router but because you are not keying your chart sub-components, so they are getting destroyed and recreated. As mentioned in https://github.com/leeoniya/domvm/issues/74#issuecomment-219231987, this will be solved by either keying the charts or creating them outside of render (making them persistent) and passing them in.

```js
return ['div',
	['div', [Chart, { interval: 2500, line_width: 3 },  "a"]],
	['div', [Chart, { interval: 5000, line_width: 8 },  "b"]],
	['div', [Chart, { interval: 60000, line_width: 2 }, "c"]],
	['div', [Chart, { interval: 20000, line_width: 1 }, "d"]]
]
```

Again, more info would be useful and allow me to help you better.

**Comment by talmobi** [comment]
No you're right -- the only reason I had router.refresh() there is because I copied it from the router module sample thinking it was mandatory (https://github.com/leeoniya/domvm#route-module).

Having insert and unmount hooks is necessary though since the init of the charts requires them to be in the DOM. So everything works perfectly if only I would add unmount hooks to the charts (since then the charts could release their requestAnimationFrame loops and resize eventlisteners).
